omegaconf/grammar/OmegaConfGrammarLexer.g4

```diff
@@ -16,13 +16,23 @@ fragment ESC_BACKSLASH: '\\\\';  // escaped backslash
 
 TOP_INTER_OPEN: INTER_OPEN -> type(INTER_OPEN), pushMode(INTERPOLATION_MODE);
 
-ESC_INTER: '\\${';
-TOP_ESC: ESC_BACKSLASH+ -> type(ESC);
+// Regular string: anything that does not contain any $ and does not end with \
+// (this ensures this rule will not consume characters required to recognize other tokens).
+ANY_STR: ~[$]* ~[\\$];
```

Comment:
`<rant>`See how supporting `$` in keys was not the best idea?  a lot of the complexity here is related to it`</rant>`

Reply:
Hehe, actually this has nothing to do with keys, this is because of the dollar sign in `INTER_OPEN` == `${`

Reply:
Oh, so this stops ANY_STR from consuming `INTER_OPEN: '${'`?
How come it's not an issue for any of the higher level constructs like lists and dicts?

Reply:
> Oh, so this stops ANY_STR from consuming `INTER_OPEN: '${'`?

Yes, it forces it to stop before the `$` sign.
Then when looking at candidates for the next token, the lexer will have to choose betwen `TOP_INTER_OPEN` and `DOLLAR`, and since the former matches more characters it will be picked.

> How come it's not an issue for any of the higher level constructs like lists and dicts?

Not sure exactly what problematic case you have in mind, but in unquoted strings we have:
```
UNQUOTED_CHAR: [/\-\\+.$%*@?];  // other characters allowed in unquoted strings
```
and importantly, this matches a single character (there is no `+`), so in an unquoted string like `@${foo}`, the lexer will first match `@`, then pick `INTER_OPEN` over `UNQUOTED_CHAR` for the next token.

I suspect this wasn't your question though... maybe a better answer is that lists and dicts are parsed in `VALUE_MODE`?

Reply:
Gotcha.